File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        map = collections.defaultdict(list)
        for word in strs:
            charCount = [0]*26
            for char in word:
                charCount[ord(char) - ord("a")] += 1
            map[tuple(charCount)].append(word)
        return map.values()
            

File: 0049-group-anagrams/0049-group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    const map = new Map();
    
    for(let word of strs){
        let charCounts = getCharCount(word);
        let values = map.get(charCounts) || [];
        values.push(word);
        map.set(charCounts, values);
    }
    
    return [...map.values()];
};

const getCharCount = word => {
    const charCount = new Array(26).fill(0);
    for(let char of word){
        let charCode = getCode(char);
        charCount[charCode]++;
    }
    return buildKey(charCount);
}

const getCode = char => char.charCodeAt(0) - "a".charCodeAt(0);

const buildKey = charCount => charCount.toString();
